contracts/smart-account/modules/AccountRecoveryModule.sol

```diff
@@ -517,6 +519,164 @@ contract AccountRecoveryModule is
+            (uint256(reqValidAfter) << (160 + 48));
+    }
+
+    /**
+     * @dev Checks if the request to be added is valid
+     * The valid request should use SA.execute => this.executeRecovery flow
+     * @param innerCallData callData of the request
+     */
+    function _validateRequestToBeAdded(
+        bytes memory innerCallData
+    ) internal view {
+        bytes4 expectedExecuteSelector;
+        address expectedThisAddress;
+        assembly {
+            //32(memory bytes array length) + 4(submitRecoveryRequest selector) + 32(offset) + 32(length)
+            expectedExecuteSelector := mload(add(innerCallData, 0x64))
+            expectedThisAddress := mload(add(innerCallData, 0x68))
+        }
+        if (
+            expectedExecuteSelector != EXECUTE_SELECTOR &&
+            expectedExecuteSelector != EXECUTE_OPTIMIZED_SELECTOR
+        ) {
+            revert("AccRecovery: WRR01"); //Wrong Recovery Request 01 = wrong execute selector in the request
+        }
+        if (expectedThisAddress != address(this)) {
+            revert("AccRecovery: WRR02"); //Wrong Recovery Request 02 = call should be to this contract
+        }
+        bytes4 expectedExecuteRecoverySelector;
+        uint256 executeRecoveryCallDataOffset;
+
+        assembly {
+            executeRecoveryCallDataOffset := mload(add(innerCallData, 0xa8))
+            expectedExecuteRecoverySelector := mload(
+                add(
+                    innerCallData,
+                    add(
+                        //executeRecovery callData start position
+                        add(
+                            0x68, //position where execute() arguments start
+                            executeRecoveryCallDataOffset //offset where executeRecovery callData bytes array start
+                        ),
+                        //skip length
+                        0x20
+                    )
+                )
+            )
+        }
+        if (expectedExecuteRecoverySelector != this.executeRecovery.selector) {
+            revert("AccRecovery: WRR03"); //Wrong Recovery Request 03 = wrong executeRecovery selector in the request
+        }
+    }
+
+    /**
+     * @dev Validates guardians signatures
+     * @param smartAccount Smart Account being recovered
+     * @param moduleSignature cleaned signature
+     * @param userOpHash hash of the userOp
+     * @return validation data (sig validation result + validUntil + validAfter)
+     */
+    function _validateGuardiansSignatures(
+        address smartAccount,
+        bytes calldata moduleSignature,
+        bytes32 userOpHash
+    ) internal view returns (uint256) {
+        uint256 requiredSignatures = _smartAccountSettings[smartAccount]
+            .recoveryThreshold;
+        if (requiredSignatures == 0) revert("AccRecovery: Threshold not set");
+
+        require(
+            moduleSignature.length >= requiredSignatures * 2 * 65,
```

Comment:
why * 2 ?

Reply:
because `moduleSignature` contains signatures over `userOpHash` and signatures over `CONTROL_MESSAGE`.
latter is used to recover guardian's address and compare it with the address that signed `userOpHash`.